README.md
# Quickly generate the specified xlsx file in Laravel

The reason for the development of this extension is that the business will often generate data in a specific XLSX format export, according to the daily development of some requirements features.integration of general functions, to achieve the requirements through the minimum configuration, do not need to see the corresponding PhpSpreadsheet package documents.However, for some complex requirements, you still need to check the phpspreadsheet documentation.


## The environment
   
```php
php >7.1
Laravel  6.*
phpoffice/phpspreadsheet ^1.11
```

## Installation
```bash
composer require remember/laravel-spreadsheet
```
 **If there is a lack of extension, please install the extension first**

Publish configuration
```bash
php artisan vendor:publish --provider="Remember\LaravelPhpSpreadsheet\LaravelPhpSpreadsheetServiceProvider"
```

#### configuration file 
config/laravel-phpSpreadsheet.php

```php
/*
 * You can place your custom package configuration in here.
 */
return [
    'startRow' => 2,
    'startLine' => 1,
    //Generate the cell format
    'columns' => [
        //demo
//        'student' => [
//            'fileName' => 'demo123'.time(),
//            'title' => 'demo1',
//            'lineName' => [
//                'name', 'age', 'address',
//            ],
//
//        ],
    ],
    'style' => [
        'format' => [
            //Default horizontal vertical center
            'alignment' => [
                /**
                 * you can set: left、right、 center、 centerContinuous、 justify、 fill
                 */
                'horizontal' => 'center',
                'vertical' => 'center',
            ],
        ],
        //default cell width
        'width' => 15,
        'extra-width' => [
            'A' => 20,
            'B' => 25,
        ],
        //style cell border
        'border' => [
            'A1:B1' => [
                'borders' => [
                    'outline' => [
                        // Border style
                        // borderStyle you can set:
                        // dashDot dashDotDot dashed dotted double hair medium';
                        // 'mediumDashDot mediumDashDotDot mediumDashed slantDashDot thick thin';
                        'borderStyle' => 'thick',
                        'color' => ['argb' => 'FFFF0000'],
                    ],
                ],

            ],

        ]
    ],

];
```
## Usage
 if you want  the browser to download the data and save it as an Excel file
```php
$data = [
            ['wuqinqiang', 23, 'HangZhou'],
            ['curry', 25, 'ShangHai'],
            ['Tom', 30, 'ShenZhen']
        ];
app('laravel-phpSpreadsheet')->downloadFile('student', $data);
```
If you want to save the file,then return file path 

```php
$path=app('laravel-phpSpreadsheet')->saveFile('student', $data);
return $path;
```
If your data is taken from the database and the data is not easily populated into the specified column.You can just generate one shelf,then populate it yourself.
                                                                                                 
```php
$data = [
            ['name'=>'wuqinqiang', 'age'=>24, 'address'=>'QuZhou'],
            ['name'=>'zhangsan', 'age'=>26, 'address'=>'HangZhou',],
            ['name'=>'lisi', 'age'=>30, 'address'=>'ShenZhen'],
        ];
        $phpSpreadsheet=app('laravel-phpSpreadsheet');
        $sheet= $phpSpreadsheet->init('student');
        $sheet = app('laravel-phpSpreadsheet')->init('student');
        $count=3;
        collect($data)->map(function ($res) use ($sheet, &$count) {
            $sheet->setCellValue('A' . $count, $res['name']);
            $sheet->setCellValue('B' . $count, $res['age']);
            $sheet->setCellValue('C' . $count, $res['address']);
            $count++;
        });
        // save file
       $file= $phpSpreadsheet->saveLocal('student');
       // download
      $phpSpreadsheet->downForBrowser('student');
```


### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email wuqinqiang050@gmail.com instead of using the issue tracker.

## Credits

- [wuqinqiang](https://github.com/wuqinqiang)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `laravel-phpspreadsheet` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


tests/ExampleTest.php
<?php

namespace Remember\LaravelPhpSpreadsheet\Tests;

use Orchestra\Testbench\TestCase;
use Remember\LaravelPhpSpreadsheet\LaravelPhpSpreadsheetServiceProvider;

class ExampleTest extends TestCase
{

    protected function getPackageProviders($app)
    {
        return [LaravelPhpSpreadsheetServiceProvider::class];
    }
    
    /** @test */
    public function true_is_true()
    {
        $this->assertTrue(true);
    }

    public function testIstrue()
    {
        return true;
    }
}


src/LaravelPhpSpreadsheet.php
<?php

namespace Remember\LaravelPhpSpreadsheet;

use Illuminate\Support\Facades\Storage;
use PhpOffice\PhpSpreadsheet\IOFactory;
use PhpOffice\PhpSpreadsheet\Spreadsheet;
use PhpOffice\PhpSpreadsheet\Writer\Xlsx;

class LaravelPhpSpreadsheet
{
    protected $spreadsheet;

    public function __construct(Spreadsheet $spreadsheet)
    {
        $this->spreadsheet = $spreadsheet;
    }

    /**
     * @param string $param
     * @return bool
     */
    public function hasParam(string $param): bool
    {
        return array_key_exists($param, config('laravel-phpSpreadsheet.columns'));
    }

    /**
     * @param $param
     * @return bool
     * @throws InvalidArgumentException
     * check the necessary configuration parameters
     */
    public function checkConfig($param): bool
    {
        if (false === $this->hasParam($param)) {
            throw new InvalidArgumentException("the $param is not in columns");
        }
        $parameters = config("laravel-phpSpreadsheet.columns.$param");
        if (!isset($parameters['fileName']) || !isset($parameters['title']) || !isset($parameters['lineName'])) {
            throw new InvalidArgumentException("please check $param  configuration item ");
        }
        if (!is_array($parameters['lineName'])) {
            throw new InvalidArgumentException('lineName must to be an array');
        }
        return true;
    }


    public function makeSheet(string $param, array $data)
    {
        $this->checkConfig($param);
        $sheet = $this->init($param);
        $row = $this->getStartRow() + 1;
        collect($data)->map(function ($res) use (&$row, &$sheet) {
            for ($i = 0; $i < count($res); $i++) {
                $sheet->setCellValueByColumnAndRow($this->getStartLine() + $i, $row, $res[$i]);
            }
            $row++;
        });
        return $sheet;
    }

    /**
     * @param $param
     * @return \PhpOffice\PhpSpreadsheet\Worksheet\Worksheet
     * @throws \PhpOffice\PhpSpreadsheet\Exception
     * initialize cell style ,setTitle ,default width
     */
    public function init($param)
    {
        $this->spreadsheet->getDefaultStyle()->applyFromArray($this->alignmentConfig());
        $sheet = $this->spreadsheet->getActiveSheet();
        $this->setStyle($sheet);
        $sheet->setTitle(config("laravel-phpSpreadsheet.columns.$param.title"));
        $sheet->getDefaultColumnDimension()->setWidth(config("laravel-phpSpreadsheet.style.width") ?? 10);
        $this->setExtraWidth($sheet);

        $startRow = $this->getStartRow();
        $startLine = $this->getStartLine();
        $lineNames = config("laravel-phpSpreadsheet.columns.$param.lineName");

        collect($lineNames)->map(function ($name) use (&$sheet, &$startRow, &$startLine) {
            $sheet->setCellValueByColumnAndRow($startLine, $startRow, $name);
            $startLine++;
        });
        return $sheet;
    }


    /**
     * @param $sheet
     * set default style
     */
    public function setStyle($sheet)
    {
        $cells = $this->getDefaultStyle();
        if ($cells && count($cells) > 0) {
            foreach ($cells as $key => $cell) {
                $sheet->getStyle($key)->applyFromArray($cell);
            }
        }
        return $sheet;
    }


    /**
     * save the file then return file url
     */
    public function saveFile($param, array $data)
    {
        $this->makeSheet($param, $data);
        return $this->saveLocal($param);
    }


    /**
     * @return mixed
     * @throws \PhpOffice\PhpSpreadsheet\Writer\Exception
     */
    public function saveLocal($param)
    {
        $writer = new Xlsx($this->spreadsheet);
        $file_name = $this->getFileName($param);
        $file = config('filesystems.disks.public.root') . '/' . $file_name;
        $writer->save($file);
        $this->disconnect();
        return Storage::disk('public')->url($file_name);
    }

    /**
     * download file
     */
    public function downloadFile($param, array $data)
    {
        $this->makeSheet($param, $data);
        $this->downForBrowser($param);
    }

    /**
     * @param $param
     * @throws \PhpOffice\PhpSpreadsheet\Writer\Exception
     */
    public function downForBrowser($param)
    {
        $file_name = $this->getFileName($param) . ".xlsx";
        header('Content-Type: application/vnd.ms-excel');
        header('Content-Disposition: attachment;filename="' . $file_name . '"');
        header('Cache-Control: max-age=0');
        $writer = IOFactory::createWriter($this->spreadsheet, 'Xlsx');
        $writer->save('php://output');
        $this->disconnect();
        exit();
    }


    /**
     * 清除
     */
    public function disconnect()
    {
        $this->spreadsheet->disconnectWorksheets();
        unset($this->spreadsheet);
    }

    /**
     * set some column width
     */
    public function setExtraWidth($sheet)
    {
        $extras = config('laravel-phpSpreadsheet.style.extra-width');
        if ($extras && count($extras) > 0) {
            foreach ($extras as $key => $extra) {
                if (strlen($key) == 1 && preg_match("/^[A-Z]+$/", $key)) {
                    $sheet->getColumnDimension($key)->setWidth($extra);
                }
            }
        }
        return $sheet;
    }

    /**
     * @return array
     * Horizontal and vertical are position
     */
    public function alignmentConfig(): array
    {
        return config('laravel-phpSpreadsheet.style.format') ?? [];
    }


    /**
     * @return int
     * get start line
     */
    public function getStartLine(): int
    {
        return config("laravel-phpSpreadsheet.startLine") > 0 ? config("laravel-phpSpreadsheet.startLine") : 1;
    }

    /**
     * @param $params
     * @return string
     * get file name from config
     */
    public function getFileName($param): string
    {
        return config("laravel-phpSpreadsheet.columns.$param.fileName") . ".xlsx";
    }

    /**
     * @return int
     * get start column
     */
    public function getStartRow(): int
    {
        return config("laravel-phpSpreadsheet.startRow") > 0 ? config("laravel-phpSpreadsheet.startRow") : 1;
    }


    /**
     * @return array
     * Styling cell borders 样式单元格边框
     */
    public function getDefaultStyle(): array
    {
        return config('laravel-phpSpreadsheet.style.border') ?? [];
    }

}


src/LaravelPhpSpreadsheetServiceProvider.php
<?php

namespace Remember\LaravelPhpSpreadsheet;

use Illuminate\Support\ServiceProvider;
use PhpOffice\PhpSpreadsheet\Spreadsheet;

class LaravelPhpSpreadsheetServiceProvider extends ServiceProvider
{

    protected $defer = true;

    /**
     * Bootstrap the application services.
     */
    public function boot()
    {
        $this->publishes([
            __DIR__ . '/../config/config.php' => config_path('laravel-phpSpreadsheet.php'),
        ], 'config');
    }

    /**
     * Register the application services.
     */
    public function register()
    {
        $this->app->singleton('laravel-phpSpreadsheet', function () {
            return new LaravelPhpSpreadsheet(new Spreadsheet());
        });
    }

    public function provides()
    {
        return [LaravelPhpSpreadsheet::class, 'xls'];
    }
}


config/config.php
<?php

/*
 * You can place your custom package configuration in here.
 */
return [
    'startRow' => 2,
    'startLine' => 1,
    //Generate the cell format
    'columns' => [
        'student' => [
            'fileName' => 'demo123' . time(),
            'title' => 'demo1',
            'lineName' => [
                'name', 'age', 'address',
            ],

        ],
    ],
    'style' => [
        'format' => [
            //Default horizontal vertical center
            'alignment' => [
                /**
                 * you can set: left right center centerContinuous justify fill
                 */
                'horizontal' => 'center',
                'vertical' => 'center',
            ],
        ],
        //default cell width
        'width' => 15,
        'extra-width' => [
            'A' => 20,
            'B' => 25,
        ],
        //style cell border
        'border' => [
            'A1:B1' => [
                'borders' => [
                    'outline' => [
                        // Border style
                        // borderStyle you can set:
                        // dashDot dashDotDot dashed dotted double hair medium';
                        // 'mediumDashDot mediumDashDotDot mediumDashed slantDashDot thick thin';
                        'borderStyle' => 'thick',
                        'color' => ['argb' => 'FFFF0000'],
                    ],
                ],

            ],
        ]
    ],

];